cmd/perflock/creds.go

```diff
@@ -5,61 +5,15 @@
 package main
 
 import (
-	"fmt"
 	"net"
-	"os"
 	"syscall"
 )
 
-// TODO: Use SO_PEERCRED instead?
-
-func writeCredentials(c *net.UnixConn) error {
-	ucred := syscall.Ucred{Pid: int32(os.Getpid()), Uid: uint32(os.Getuid()), Gid: uint32(os.Getgid())}
-	credOob := syscall.UnixCredentials(&ucred)
-	credMsg := []byte("x")
-	n, oobn, err := c.WriteMsgUnix(credMsg, credOob, nil)
-	if err != nil {
-		return err
-	}
-	if n != 1 {
-		return fmt.Errorf("short send (%d bytes)", n)
-	}
-	if oobn != len(credOob) {
-		return fmt.Errorf("short OOB send (%d bytes)", oobn)
-	}
-	return nil
-}
-
 func readCredentials(c *net.UnixConn) (*syscall.Ucred, error) {
-	// Enable receiving credentials on c.
 	f, err := c.File()
 	if err != nil {
 		return nil, err
 	}
-	err = syscall.SetsockoptInt(int(f.Fd()), syscall.SOL_SOCKET, syscall.SO_PASSCRED, 1)
-	f.Close()
-	if err != nil {
-		return nil, err
-	}
-
-	// Receive credentials.
-	buf := make([]byte, 1)
-	oob := make([]byte, 128)
-	n, oobn, _, _, err := c.ReadMsgUnix(buf, oob)
-	if err != nil {
-		return nil, err
-	}
-	if n != 1 {
-		return nil, fmt.Errorf("expected 1 byte, got %d", n)
-	}
-
-	// Parse OOB data.
-	scms, err := syscall.ParseSocketControlMessage(oob[:oobn])
-	if err != nil {
-		return nil, err
-	}
-	if len(scms) != 1 {
-		return nil, fmt.Errorf("expected 1 control message, got %d", len(scms))
-	}
-	return syscall.ParseUnixCredentials(&scms[0])
+	defer f.Close()
+	return syscall.GetsockoptUcred(int(f.Fd()), syscall.SOL_SOCKET, syscall.SO_PEERCRED)
```

Comment:
I believe people use perflock on macOS (just without the governor stuff). Does SO_PEERCRED work there?

Reply:
Huh, it seems less useful to me without being able to affect the CPU governor, disabling it was decisive in  correcting a misleading benchmark lately.

At any rate, I looked around and I think `SO_PEERCRED` may not work, but macOS appears to have something very similar called `LOCAL_PEERCRED`:

- Example: https://lists.gnu.org/archive/html/guix-commits/2015-01/msg01008.html
- macOS offers a libc function called [getpeereid(3)](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man3/getpeereid.3.html) apparently built on top of it.

Reply:
I found https://github.com/golang/go/issues/27613, which lead me to https://pkg.go.dev/inet.af/peercred, which is basically exactly what we need. I don't know how you feel about dependencies, but this is a tiny one, and it does basically exactly what I'd do, but more completely. I've added another commit using it. PTAL.